service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java

```diff
@@ -92,11 +114,107 @@ protected void extractRequiredJsonPayload(final RoutingContext ctx) {
                     ctx.response().setStatusMessage("Empty body");
                     ctx.fail(HttpURLConnection.HTTP_BAD_REQUEST);
                 }
-            } catch (DecodeException e) {
+            } catch (final DecodeException e) {
                 ctx.response().setStatusMessage("Invalid JSON");
                 ctx.fail(HttpURLConnection.HTTP_BAD_REQUEST);
             }
         }
     }
 
+    /**
+     * Set the body of the response for the HTTP request as JSON. It is retrieved from the passed in JSON result (received as answer from a request to the event bus) by
+     * getting the key {@link RequestResponseApiConstants#FIELD_PAYLOAD}.
+     *
+     * @param jsonResult The JSON result returned from the processing event bus consumer.
+     * @param response The HTTP response to which the body is written.
+     */
+    protected void setResponseBody(final JsonObject jsonResult, final HttpServerResponse response) {
+        final JsonObject msg = jsonResult.getJsonObject(RequestResponseApiConstants.FIELD_PAYLOAD);
+        if (msg != null) {
+            final String body = msg.encodePrettily();
+            response.putHeader(HttpHeaders.CONTENT_TYPE, HttpUtils.CONTENT_TYPE_JSON_UFT8)
+                    .putHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(body.length()))
+                    .write(body);
+        }
+    }
+
+    /**
+     * Get a response handler that implements the default behaviour for responding to the HTTP request.
+     *
+     * @param ctx The routing context of the request.
+     * @param action The action of the HTTP request that is being answered.
+     * @param setResponseBodyForStatus A Predicate that defines for which status a response body shall be set. Maybe null.
+     * @param locationFormatter A Function that constructs the String being set as the HTTP location header {@link HttpHeaders#LOCATION}.
+     *                          Must not be null if the HTTP request was to add an object.
+     * @return BiConsumer<Integer, JsonObject> A consumer for the status and the JSON object that implements the default behaviour for responding to the HTTP request.
+     * @throws NullPointerException If ctx is null, if the locationFormatter is null although the request was for adding an object.
+     */
+    protected final BiConsumer<Integer, JsonObject> getDefaultResponseHandler(
+            final RoutingContext ctx,
+            final RequestResponseApiConstants.Action action,
+            final Predicate<Integer> setResponseBodyForStatus,
+            final Function<Void,String> locationFormatter
```

Comment:
I would instead use a `Handler<HttpServerResponse>` and let the handler modify the response in any way it sees fit, i.e. not limiting it to provide a location header value.

Reply:
:+1: much better indeed.
And this is solving another issue as well: the annoying action parameter, which was rather artificial in this signature.